### ISSUE TITLE
Breaking change in flask-sqlalchmey 3.1.0

### ISSUE THREAD

**Issue by johnjiang** [pallets-eco/flask-sqlalchemy#1266]
In the [release notes](https://flask-sqlalchemy.palletsprojects.com/en/3.1.x/changes/#version-3-1-0) it states:

> Bump minimum version of SQLAlchemy to 2.0.16.

However, this is a breaking change, as it means that folks that are still on 1.x cannot upgrade to >=3.1.0

Similarly dropping support for 3.7 should technically be a breaking change and warrant a major bump as well.

**Comment by ThiefMaster** [comment]
If you use a tool to manage your (transitive) dependencies then it will not install incompatible combinations. There is no need to rely on semver for this.

> Similarly dropping support for 3.7 should technically be a breaking change and warrant a major bump as well.

No, that makes no sense in Python. When dropping support for a Python version, you set the `python_requires` metadata accordingly, so if you are on a now-unsupported python versions you will simply not get the incompatible versions.

In fact, even in the JS world it's incredibly annoying to get semver-major changes for dropping old environments - semver-major means I now have to check if there are any incompatibilities, even though it's almost never the case if there aren't any other (real) breaking changes.

**Comment by ThiefMaster** [comment]
> I'm using poetry and it's refusing to upgrade flask-sqlalchmey because it returns 2.x while i'm still on 1.x.

That sounds like it's doing the right thing, isn't it? It gives you the latest version that's compatible with the rest of your packages...

**Comment by johnjiang** [comment]
Yes it is, but I think you're still not understanding me. If I can install flask-sqlalchmey 3.0 on my project, installing 3.1 should still be possible

**Comment by ThiefMaster** [comment]
From a "semver-puristic" view: Yes.

But otherwise: Why? A bugfix for some important bug in 3.0 could be released as 3.0.x.

There won't be new features in 3.0, so there's no need to keep the 3.x version numbers available for new feature releases (once again from a semver point of view since there's nothing preventing a new feature in a 3.0.x release).

**Comment by davidism** [comment]
There is no requirement that A.B.C corresponds to semver's ideas of those values. In Pallets, like Python itself, you can think of the numbers as milestone.feature.fix instead. It's perfectly possible to use Poetry in that situation, although it does unfortunately assume everything is major.minor.patch by default and so gives you the wrong constraints.

We will not be changing how we increment versions or what version parts mean. Please review any of the following for more information about not assuming semver for anything:

* https://hynek.me/articles/semver-will-not-save-you/
* https://www.youtube.com/watch?v=WSVFw-3ssXM&t
* https://snarky.ca/why-i-dont-like-semver/
* https://caremad.io/posts/2016/02/versioning-software/
* https://bernat.tech/posts/version-numbers/
* https://iscinumpy.dev/post/bound-version-constraints/

**Comment by johnjiang** [comment]
I feel like we're splitting hairs over this. All I'm saying is that we've added extra unnecessary burden on the community by introducing a backwards breaking change in a minor upgrade. If I encountered this issue, I'm sure many more engineers would as well. You might not feel this is an important issue and that's your prerogative, all I'm saying is that it doesn't hurt to bump up to 4.0 to make it more explicit that we're supporting sqlalchemy 1.x in flask-sqlalchemy 3.x, and sqlalchmey 2.x in flask-sqlalchmey 4.x.

I can understand that we chose not to follow semver, that's fine. We can agree to disagree.

I do appreciate your continuing support flask-sqlalchemy nevertheless.
